src/language/class-modifiers.md

```diff
@@ -0,0 +1,240 @@
+- Ensuring that adding new members to the class won't risk breaking changes for 
+its consumers.
+- Ensuring access to private members on instances of the final class will never
+throw a [`noSuchMethod`][] error.
+
+## `sealed`
+
+Use the `sealed` modifier to define a class that **cannot be constructed, extended,
+or implementated** outside of the library where the class is defined. 
+
+The purpose of the `sealed` modifier is to enable [_exhaustiveness checking_][]
+for switch cases.
+Any possible direct subtypes of a sealed type can only exist in the same library,
+so the compiler is always aware of the extent of a sealed type's hierarchy.
+
+This allows the compiler to alert you when a switch is not exhaustively handling
+all possible subtypes in its cases.
+
+```dart
+sealed class Vehicle {
+  String make;
+  String model;
+  void moveForward(int meters) { … }
+}
+
+class Car extends Vehicle { }
+class Truck implements Vehicle { }
+class Bicycle extends Vehicle { }
+
+// ...
+
+return switch (vehicle) {     // ERROR: The switch is missing a subtype of Vehicle
+  Car => 'vroom',
+  Truck => 'VROOOOMM'
+};
+```
+
+## `abstract`
+
+Use the `abstract` modifier to define a class that can’t be constructed.
+Abstract classes are useful for defining interfaces, often with some
+implementation. If you want your abstract class to appear to be instantiable,
+define a [factory constructor][].
+
+Abstract classes often have [abstract methods][].
+Here’s an example of declaring an abstract class that has an abstract
+method:
+
+<?code-excerpt "misc/lib/language_tour/classes/misc.dart (abstract)"?>
+```dart
+// This class is declared abstract and thus
+// can't be instantiated.
+abstract class AbstractContainer {
+  // Define constructors, fields, methods...
+
+  void updateChildren(); // Abstract method.
+}
+```
+
+## `class`, `mixin`, or `mixin class`?
+
+The `mixin` declaration defines a [mixin][]. The `mixin` _modifier_ can only be
+applied to a `class` declaration to create a _mixin class_: a class that can be
+treated both as a class and a mixed in. Regular classes cannot be mixed in.
+
+Mixins and mixin classes are meant to be mixed in, so they cannot be combined
+with the modifiers that prevent inheritance: `interface`, `final`, and `sealed`.
+
+## Combining modifiers
+
+It is possible to combine some modifiers for layered restrictions. For example, 
+an `abstract interface class` combines restrictions on instantiation and inheritance.
+This could be useful to define a "pure contract", or a class that doesn't update
+any preexisting state.
+
+Modifiers that are mutually exclusive cannot be combined. For example,
+`base` allows extending and disallows implementing, and `interface` allows
+implementing and disallows extending.
+
+Modifiers that are redundant cannot be combined. For example, mixins and sealed
+classes already cannot be constructed, so combining them with `abstract`, which
+disallows construction, is redundant. 
```

Comment:
This way of describing it, as if all modifiers are the same and the rules for how they can be combined are complicated, gets me very confused about something that's fairly simple (to me).

A class declaration can be, in order:
* Optionally an `abstract`, describing whether the class can contain abstract members (and it also prevents instantiation).
* Optionally one of `base`, `interface`, `final` or `sealed`, describing restrictions on other libraries subtyping the class.
* Optionally a `mixin`, describing whether the declaration also introduces a mixin (aka. can be mixed in).
* and then the `class` keyword itself.

Some combinations are disallowed:
* `abstract` with `sealed` (because a `sealed` class is always implicitly `abstract`).
* `interface`, `final` or `sealed` with `mixin` (because those access modifiers prevent mixing in).

The phrasing above doesn't say that `abstract` has to come first, and it sounds like combining `abstract` and `base` is at the same level as combining `base` and `interface`. To me it isn't. There are different groups of modifiers that can be combined (almost) orthogonally.

Reply:
That's a lot clearer, I think I'll rewrite the entire section to more or less match this comment